Update Docs with Finalize Block (#760)
cometbft/cometbft: docs/guides/go.md

Finally, make sure to add the bytes package to the `import` stanza at the top of `app.go`:

```go
import(
    "bytes"

    "github.com/dgraph-io/badger/v3"
    abcitypes "github.com/cometbft/cometbft/abci/types"
)
```

### 1.3.3 FinalizeBlock

When the CometBFT consensus engine has decided on the block, the block is transferred to the
application via the `FinalizeBlock` method.
`FinalizeBlock` is an ABCI method introduced in CometBFT `v0.38.0`. This replaces the functionality provided previously (pre-`v0.38.0`) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.
`FinalizeBlock`'s parameters are an aggregation of those in `BeginBlock`, `DeliverTx`, and `EndBlock`.

This method is responsible for executing the block and returning a response to the consensus engine.
Providing a single `FinalizeBlock` method to signal the finalization of a block simplifies the ABCI interface and increases flexibility in the execution pipeline.

The `FinalizeBlock` method executes the block, including any necessary transaction processing and state updates, and returns a `ResponseFinalizeBlock` object which contains any necessary information about the executed block.

> **Note:** `FinalizeBlock` only prepares the update to be made and does not change the state of the application. The state change is actually committed in a later stage i.e. in `commit` phase..

Reviewer: ```suggestion
**Note:** `FinalizeBlock` only prepares the update to be made and does not change the state of the application. The state change is actually committed in a later stage i.e. in `commit` phase.
```

```go
func (app *KVStoreApplication) FinalizeBlock(_ context.Context, req *abcitypes.RequestFinalizeBlock) (*abcitypes.ResponseFinalizeBlock, error) {
    var txs = make([]*abcitypes.ExecTxResult, len(req.Txs))

    app.onGoingBlock = app.db.NewTransaction(true)
    for i, tx := range req.Txs {
        if code := app.isValid(tx); code != 0 {
            log.Printf("Error in tx in if")
            txs[i] = &abcitypes.ExecTxResult{Code: code}
        } else {
            parts := bytes.SplitN(tx, []byte("="), 2)
            key, value := parts[0], parts[1]
            log.Printf("Adding key %s with value %s", key, value)

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }
            log.Printf("Successfully added key %s with value %s", key, value)

            txs[i] = &abcitypes.ExecTxResult{}
        }
    }

    return &abcitypes.ResponseFinalizeBlock{
        TxResults:        txs,
    }, nil
}
```

Transactions are not guaranteed to be valid when they are delivered to an application, even if they were valid when they were proposed.

> This can happen if the application state is used to determine transaction validity. The application state may have changed between the initial execution of CheckTx and the transaction delivery in FinalizeBlock in a way that rendered the transaction no longer valid.

Reviewer: ```suggestion
This can happen if the application state is used to determine transaction validity. The application state may have changed between the initial execution of `CheckTx` and the transaction delivery in `FinalizeBlock` in a way that rendered the transaction no longer valid.
```

> **Note** that FinalizeBlock cannot yet commit the Badger transaction we were building during the block execution.

Reviewer: ```suggestion
**Note** that `FinalizeBlock` cannot yet commit the Badger transaction we were building during the block execution.
```

> Other methods, such as Query, rely on a consistent view of the application's state, the application should only update its state by committing the Badger transactions when the full block has been delivered and the Commit method is invoked.

Reviewer: ```suggestion
Other methods, such as `Query`, rely on a consistent view of the application's state, the application should only update its state by committing the Badger transactions when the full block has been delivered and the `Commit` method is invoked.
```

Note that `CheckTx` does not execute the transaction, it only verifies that that the transaction could be executed. We do not know yet if the rest of the network has agreed to accept this transaction into a block.

Finally, make sure to add the bytes package to the `import` stanza at the top of `app.go`:

```go
import(
    "bytes"

    "github.com/dgraph-io/badger/v3"
    abcitypes "github.com/cometbft/cometbft/abci/types"
)
```

### 1.3.3 FinalizeBlock

> `FinalizeBlock` is a method introduced in CometBFT v0.38.0 which replaces the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock`.

Reviewer 2: ```suggestion
```

### 1.3.3 FinalizeBlock

### 1.3.3 BeginBlock -> DeliverTx -> EndBlock -> Commit

When the CometBFT consensus engine has decided on the block, the block is transferred to the
application over three ABCI method calls: `BeginBlock`, `DeliverTx`, and `EndBlock`.

- `BeginBlock` is called once to indicate to the application that it is about to
receive a block.
- `DeliverTx` is called repeatedly, once for each application transaction that was included in the block.
- `EndBlock` is called once to indicate to the application that no more transactions
will be delivered to the application in within this block.
`FinalizeBlock` is a method introduced in CometBFT v0.38.0 which replaces the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock`.

Note that, to implement these calls in our application we're going to make use of Badger's
transaction mechanism. We will always refer to these as Badger transactions, not to
confuse them with the transactions included in the blocks delivered by CometBFT,

> the _application transactions_.

Reviewer: This paragraph is important not to confuse the reader with the two different meanings of the term "transaction" (database transactions vs. blockchain transactions). Is it kept somewhere?

> This method is responsible for executing the block and returning a response to the consensus engine. By implementing `ProcessProposal`, transactions are processed before the block execution phase and are already present in the state machine. This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. This simplifies the ABCI interface and increases flexibility in the execution pipeline.

Reviewer: As in `go-built-in`, mentioning immediate execution here seems to advanced to me for a reader that just had the first contact with the interface. Remember that `ProcessProposal` is not obliged to fully execute the proposed block: it is optional (only advanced apps will do it)

>         if code := app.isValid(tx); code != 0 {

Reviewer 2: same comments as for the in-processes guide.

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }
            log.Printf("Successfully added key %s with value %s", key, value)

            // Marshal the ExecTxResult to the txs slice element
            txs[i] = &abcitypes.ExecTxResult{}

>             if _, err := txs[i].MarshalToSizedBuffer(nil); err != nil {

Reviewer: Why do we need to do this check here? If we need to serialize and it fails, we'll panic at that point

                log.Panicf("Error marshalling tx result: %v", err)
            }
        }
    }

    return &abcitypes.ResponseFinalizeBlock{
        TxResults:        txs,

>         AppHash:          []byte{},
>         Events:           []abcitypes.Event{},
>         ValidatorUpdates: abcitypes.ValidatorUpdates{},

Reviewer: ```suggestion
```

            txs[i] = &abcitypes.ExecTxResult{Code: code}
        } else {
            parts := bytes.SplitN(tx, []byte("="), 2)
            key, value := parts[0], parts[1]
            log.Printf("Adding key %s with value %s", key, value)

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }
            log.Printf("Successfully added key %s with value %s", key, value)

            // Marshal the ExecTxResult to the txs slice element
            txs[i] = &abcitypes.ExecTxResult{}
            if _, err := txs[i].MarshalToSizedBuffer(nil); err != nil {
                log.Panicf("Error marshalling tx result: %v", err)
            }
        }
    }

    return &abcitypes.ResponseFinalizeBlock{
        TxResults:        txs,
        AppHash:          []byte{},
        Events:           []abcitypes.Event{},
        ValidatorUpdates: abcitypes.ValidatorUpdates{},
    }, nil
}
```

Next, let's modify `DeliverTx` to add the `key` and `value` to the database transaction every time our application

> receives a new application transaction through `RequestDeliverTx`.

Reviewer: `FinalizeBlock` still does this. Is this paragraph (or some text that explains the same) somewhere?

		log.Panicf("Error writing to database, unable to execute tx: %v", err)
	}

	return abcitypes.ResponseDeliverTx{Code: 0}
}
```

Note that we check the validity of the transaction _again_ during `DeliverTx`.
Transactions are not guaranteed to be valid when they are delivered to an
application, even if they were valid when they were proposed.
This can happen if the application state is used to determine transaction
validity. Application state may have changed between the initial execution of `CheckTx`
and the transaction delivery in `DeliverTx` in a way that rendered the transaction
no longer valid.

`EndBlock` is called to inform the application that the full block has been delivered
and give the application a chance to perform any other computation needed, before the
effects of the transactions become permanent.

Note that `EndBlock` **cannot** yet commit the Badger transaction we were building
in during `DeliverTx`.
Since other methods, such as `Query`, rely on a consistent view of the application's
state, the application should only update its state by committing the Badger transactions
when the full block has been delivered and the `Commit` method is invoked.

> 

Reviewer: Same comment as for the `go-built-in` version

>     if err := app.onGoingBlock.Commit(); err != nil {
>         return &abcitypes.ResponseCommit{}, err
>     }
>     return &abcitypes.ResponseCommit{}, nil

Reviewer: ```suggestion
    return &abcitypes.ResponseCommit{}, app.onGoingBlock.Commit()
```

Same comment as in the `go-built-in` version: the non-`nill` err will cause ComtBFT to panic, but I find it easier to follow if the panic appears here explicitly (the reader does not necessarily know that CometBFT panics with non-`nil` error returned by ABCI calls).

Note that `CheckTx` does not execute the transaction, it only verifies that that the transaction could be executed. We do not know yet if the rest of the network has agreed to accept this transaction into a block.

Finally, make sure to add the bytes package to the `import` stanza at the top of `app.go`:

```go
import(
    "bytes"

    "github.com/dgraph-io/badger/v3"
    abcitypes "github.com/cometbft/cometbft/abci/types"
)
```

### 1.3.3 FinalizeBlock

> `FinalizeBlock` is an ABCI method introduced in CometBFT `v0.38.0`. This replaces the functionality provided previously (pre-`v0.38.0`) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.

Reviewer: ```suggestion
When the CometBFT consensus engine has decided on the block, the block is transferred to the
application via the `FinalizeBlock` method.
`FinalizeBlock` is an ABCI method introduced in CometBFT `v0.38.0`. This replaces the functionality provided previously (pre-`v0.38.0`) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.
`FinalizeBlock`'s parameters are an aggregation of those in `BeginBlock`, `DeliverTx`, and `EndBlock`.
```

This method is responsible for executing the block and returning a response to the consensus engine.

> The `ProcessProposal` method provides the transactions to the state machine before the block is finalized.
> This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. 

Reviewer: ```suggestion
```

Providing a single `FinalizeBlock` method to signal the finalization of a block simplifies the ABCI interface and increases flexibility in the execution pipeline.

The `FinalizeBlock` method executes the block, including any necessary transaction processing and state updates, and returns a `ResponseFinalizeBlock` object which contains any necessary information about the executed block.

> **Note:** `FinalizeBlock` only prepares the update to be made and does not change the state of the application. The change of state happens in a later stage i.e. is `commit` phase.

Reviewer: ```suggestion
**Note:** `FinalizeBlock` only prepares the update to be made and does not change the state of the application. The state change is actually committed in a later stage i.e. in `commit` phase..
```

Next, let's modify `FinalizeBlock` to add the `key` and `value` to the database transaction every time our application receives a new application transaction through `RequestFinalizeBlock`.

Note that we check the validity of the transaction _again_ during `FinalizeBlock`.

```go
func (app *KVStoreApplication) FinalizeBlock(_ context.Context, req *abcitypes.RequestFinalizeBlock) (*abcitypes.ResponseFinalizeBlock, error) {
    var txs = make([]*abcitypes.ExecTxResult, len(req.Txs))

    app.onGoingBlock = app.db.NewTransaction(true)
    for i, tx := range req.Txs {
        if code := app.isValid(tx); code != 0 {
            log.Printf("Error in tx in if")
            txs[i] = &abcitypes.ExecTxResult{Code: code}
        } else {
            parts := bytes.SplitN(tx, []byte("="), 2)
            key, value := parts[0], parts[1]
            log.Printf("Adding key %s with value %s", key, value)

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }
            log.Printf("Successfully added key %s with value %s", key, value)

>             // Marshal the ExecTxResult to the txs slice element

Reviewer: ```suggestion
```
